dask/base.py

```diff
@@ -0,0 +1,85 @@
+import warnings
+from operator import attrgetter
+
+from toolz import merge, groupby, unique
+
+from .context import _globals
+
+
+class Base(object):
+    """Base class for dask collections"""
+
+    def visualize(self, optimize_graph=False):
+        from dask.dot import dot_graph
+        if optimize_graph:
+            return dot_graph(self._config.optimize(self.dask, self._keys()))
+        else:
+            return dot_graph(self.dask)
+
+    def _visualize(self, optimize_graph=False):
+        warn = DeprecationWarning("``_visualize`` is deprecated, use "
+                                  "``visualize`` instead.")
+        warnings.warn(warn)
+        return self.visualize(optimize_graph)
+
+    def compute(self, **kwargs):
+        return compute(self, **kwargs)[0]
+
+
+class Config(object):
+    """Specifies a configuration for a dask collection.
+
+    This holds functions that are used to specialize dask schedulers for a
+    specific collection.
+
+    Parameters
+    ----------
+    optimize : callable
+        Takes a dask and iterable of keys, and returns an optimized dask.
+    default_get : callable
+        The default scheduler to use (e.g. ``dask.threaded.get``)
+    finalize : callable
+        Takes a dask collection and an iterable of results, and converts them
+        into the desired python type.
+    """
+    def __init__(self, optimize, default_get, finalize):
+        self.optimize = optimize
+        self.default_get = default_get
+        self.finalize = finalize
```

Comment:
Before, each collection had the following lines:
```python
def get(dsk, keys, get=None, **kwargs):
    get = get or _globals['get'] or some_get_function
    dsk2 = optimize(dsk, keys)
    return get(dsk2, keys, **kwargs)
```

which were used in several places throughout the collection for intermediate computations (the few times where we're not lazy). This function could be changed without also modifying the parameters attached to the collection class, which seems wrong to me. It also had to be repeated several times (one for each collections module). Using this class (even with the lines needed to write the class) results in a shorter code base, with fewer repeated lines, and less danger of changing something in one place but not the other.